File: #1578.minimum-time-to-make-rope-colorful.cpp
class Solution {
public:
    int minCost(string colors, vector<int>& neededTime) {

        int size = colors.size();
        int max = 0;
        int sum = 0;
        int result = 0;
        for (int i = 0; i < size; ++i) {
            if (i - 1 >= 0 && colors[i] != colors[i - 1]) {
                result += sum - max;
                sum = max = 0;
            }
            sum += neededTime[i];
            max = std::max(max, neededTime[i]);
        }
        result += sum - max;
        return result;
    }
};

File: #1578.minimum-time-to-make-rope-colorful.js
/**
 * @param {string} colors
 * @param {number[]} neededTime
 * @return {number}
 */
var minCost = function(colors, neededTime) {
    const length = neededTime.length
    if (length < 2) {
        return 0;
    }

    let [count, max, val] = [0, neededTime[0]];
    for (let i = 0; i < length - 1; i++) {
        if (colors[i] !== colors[i + 1]) {
             max = neededTime[i + 1];
             continue;
        }

        val = neededTime[i + 1];
        if (max > val) {
            count += val;
        } else {
            count += max;
            max = val;
        }
    }
    return count;
};
